enhancements/sig-policy/71-hub-templates-auto-reconcile/README.md

```diff
@@ -0,0 +1,232 @@
+}
+
+func New(options *Options) (*DynamicWatcher, error) {
+   ...
+}
+```
+
+The new library will deduplicate watches, meaning if two objects (e.g. a Policy) watch the same object (e.g. a
+ConfigMap), then only one Kubernetes API watch request must be made. Watch API requests must also be limited to just the
+objects being watched and not all objects in the namespace for efficiency and scale.
+
+The new library would likely use the client-go dynamic Watch method since it is high level enough where retries,
+backoffs, and unmarshaling is handled while not caching the results like a shared informer would. This is important
+because there isn’t a need to cache the watched objects, but only to know when they have changed. If there are a lot of
+watched objects, this will keep memory consumption low.
+
+The implementation will need two data structures. The first needs to map watcher objects (e.g. a Policy) to the watches
+it’s “subscribed” to. This will be helpful when updating or removing watches. The second data structure maps the watched
+objects to the watcher objects (e.g. a Policy). This will be helpful to populate the
+[client-go rate limiting queue](https://pkg.go.dev/k8s.io/client-go@v0.23.5/util/workqueue?utm_source=gopls#NewNamedRateLimitingQueue)
+when a watch notifies of an object updating. It will also be helpful to determine how many watcher objects (e.g a
+Policy) are watching this object to determine if the watch API request should be canceled after the RemoveWatcher method
+is called.
+
+#### Using The Library
+
+The Policy controller in the Governance Policy Propagator would initially be the only consumer of the library. To use
+it, the controller would essentially have two event loops. The first is the existing reconcile loop that runs on updates
+to Policy and Placement related objects. This event loop would be updated so that any time a root Policy object is
+updated that uses Hub policy templates, dynamic watches would be added, updated, or removed using the new library. To do
+this, the [go-template-utils](https://github.com/stolostron/go-template-utils) library would need to be updated to
+return the objects it queried for on the API. The reconcile loop would then collect references to all objects that were
+queried for using Hub policy templates for that root Policy object and call AddOrUpdateWatcher method from the library.
+When the root Policy object is deleted, the RemoveWatcher method from the library is called.
+
+The second (new) event loop would be the event loop from the new library, where anytime an object referenced in a Hub
+policy template is updated, the root Policy would be reconciled. This would have the same logic as the first event loop,
+and thus this new Reconcile method would just wrap the Reconcile method from the first event loop. Note that since these
+event loops will be run as different Goroutines, locking should be added so that only one root Policy object is
+reconciled at a time.
```

Comment:
Why is this locking necessary? I would understand needing to lock whatever structures are used for mapping policies <-> objects in templates, but I don't see why the whole reconcile needs to be locked.

For example, if a secret is used in two policies, I don't see why those would need to be reconciled sequentially when the secret is changed.

Reply:
The lock would be at the root policy level and not based on the referenced object.

What I want to try to avoid is two goroutines updating the same policy and causing one's update to fail because a replicated policy was updated by the other goroutine since the replicated policy was pulled from the cache. This would then cause unnecessary log noise and additional execution of code that isn't necessary.